validation/scenario_generator.py

```diff
@@ -136,14 +136,18 @@ def generate_scenario(start_date_str, end_date_str, raw_df, countries=None, scen
                 new_rows.append(new_row)
                 # Move to next day
                 current_date = current_date + np.timedelta64(1, 'D')
-    if new_rows:
-        future_rows_df = pd.DataFrame(new_rows, columns=ips_df.columns)
-        # Delete any old row that might have been replaced by a scenario one
-        replaced_dates = list(future_rows_df["Date"].unique())
-        ips_df = ips_df[ips_df["Date"].isin(replaced_dates) == False]
-        # Append the new rows
-        ips_df = ips_df.append(future_rows_df)
-        # Sort
-        ips_df.sort_values(by=ID_COLS, inplace=True)
+            # Add the new rows
+            if new_rows:
+                new_rows_df = pd.DataFrame(new_rows, columns=ips_df.columns)
+                # Delete any old row that might have been replaced by a scenario one for this country / region
+                replaced_dates = list(new_rows_df["Date"].unique())
+                rows_to_drop = ips_df[(ips_df.CountryName == country) &
```

Comment:
We were blindly removing dates for which we had generated a scenario, assuming they'd be replaced by the scenario ones. But scenario dates can be different by country, when one country has 1 more day of data than other for instance. So this logic is now done at the country / region level.